### ISSUE TITLE
spike: how to launch comm=ofi programs on non-Cray systems?

### ISSUE THREAD

**Issue by gbtitus** [chapel-lang/chapel#11116]
What should the launch mechanism be for comm=ofi Chapel programs on non-Cray systems?  On Cray X* systems we can use the same Chapel launchers we do now, and the comm layer can get setup info from PMI just as comm=ugni does.  But what should we do on other systems where we don't have PMI?

**Comment by gbtitus** [comment]
My first experiment will be to take a look at what happens with the sockets provider when a 2-node program is launched with slurm's `srun`.

**Comment by gbtitus** [comment]
A note: this is a long-term need, not a short-term one.  For near-term development we can certainly rely on Cray X* systems with PMI.  Also, this doesn't affect how the libfabric-related coding is done, except perhaps in some tiny way during initialization/setup.  It's really more an issue with how the out-of-band coding is done, notably what to do in place of `PMI_Allgather()`.

**Comment by gbtitus** [comment]
This story generated extensive conversation during sprint planning, so here is some more background information.

Comm layers may require support for a small number of so-called "out of band" operations.  These are operations needed during comm layer initialization and perhaps shutdown, when the comm layer itself is not operational and yet information needs to be exchanged among the nodes.  As an example, in order to bootstrap inter-node communication, the comm layer on each node needs the transport layer addresses of all the other nodes, but by definition it needs those addresses before it can actually communicate with those nodes. Typical operations needed are:

- get the number of nodes
- get this node's index in `[0..numNodes]`
- `barrier` across all nodes
- `all_to_all` exchange information among all nodes.

How these out-of-band operations are provided is tied fairly closely to the system or transport launch mechanism. For comm=gasnet, GASNet itself provides these operations, with the GASNet launchers doing the backing work to implement them:
- number of nodes: `gasnet_nodes()`
- node index: `gasnet_mynode()`
- `barrier`: `gasnet_barrier_notify()` and `gasnet_barrier_try()`
- `all_to_all`: implicit in the `gasnet_init()` and various GASNet segment-oriented functions.

In comm=ugni we use `PMI` (the Process Management Interface, see `man pmi` on a Cray X\* system) for out-of-band support, since that is integrated with both of the Cray system launchers:
- number of nodes: `PMI_Get_size()`
- node index: `PMI_Get_rank_in_app()`
- `barrier`: `PMI_Barrier()`
- `all_to_all`: `PMI_Allgather()`.

In `comm=ofi` we can and do use `PMI` on Cray X\* systems. But we don't yet have out-of-band support for other systems.  This story has to do with exploring what we might use for that, including exploring whether there are `libfabric` capabilities we could use to implement it within the comm layer itself. Admittedly the title "how to launch" doesn't indicate this, which was a little misleading. But at the same time there is pretty much a 1:1 relationship between the launch mechanism and the out-of-band support, such that picking one also picks the other, or at least very much constrains our options for the other.

**Comment by gbtitus** [comment]
I've been doing a little experimenting launching comm=ofi and other programs using `srun` on linux64.  It's easy to get the number of nodes and each program instance's node ID from the environment slurm sets for `srun` client programs: they are in `SLURM_NTASKS` and `SLURM_PROCID`, respectively. And since `barrier` can be done using a vacuous `all_to_all`, it's really just the latter that we're missing with a bare `srun`.

**Comment by gbtitus** [comment]
The key to making connections with other program instances in `ofi` is knowing their addresses, that is, what those other instances see when they call `fi_getname()`.  In the comm layer each program instance gets its own address that way and then all the addresses are broadcast around the program using an `all_to_all` (`chpl_comm_ofi_oob_allgather()`, in comm=ofi). This is the first and primary `all_to_all` in the ofi comm layer.  There is another one to broadcast all the registered memory regions, but that could be replaced with an AM-based technique that updated initiating nodes' knowledge of remote target node memory regions as references were made.  An `all_to_all` is also used to broadcast the data structures used in native comm layer barriers (vs. out-of-band ones) once the comm layer is running, but that could also easily be replaced with an AM-based equivalent. It's really the address exchange that is key.

With the sockets provider, the node addresses are of the form `<IP addr>:<port>`, where `<IP addr>` is the node's IP address in dotted quad format and `<port>` is an ephemeral port number, meaning one outside the set defined in `/etc/services`.  The port numbers are not the same across the nodes, that is, each node seems to choose its own port number more or less randomly.  The actual address format doesn't actually matter, but this gives an idea of the amount of data that needs exchanging; it's 15-20 bytes per node, and never more than 21 bytes per node.

It would be straightforward for the `srun` launcher to optionally provide an `all_to_all` capability. It could open a port on the launch node before doing the srun and communicate that port number to the client compute processes, say via the environment.  The launcher host's dotted-quad address is already in the slurm-provided environment, via `SLURM_LAUNCH_NODE_IPADDR`. With the launcher host IP address and port, the clients could all connect to the launcher process.  The latter could then implement `all_to_all` using a simple bulk-synchronous algorithm that read the contributed bytes from every node, wrote the concatenated result back to all of them, and repeated.

**Comment by gbtitus** [comment]
The technique I described in the last paragraph of the [comment above](https://github.com/chapel-lang/chapel/issues/11116#issuecomment-446728170) obviously doesn't scale very well. The tightest limit is probably the maximum number of open files in the launcher, since that would need an open file for each connection to a client.  Other problems at scale would be that all the clients attempting to `connect()` to it at once might overwhelm the launch node with network requests, causing timeouts or other problems.  (Slurm has a related problem; note the discussion of `PMI_TIME` in `man srun`.)

These problems could be addressed by going to a tree-based communication architecture, but at present we don't have a need to scale beyond about 16 nodes.

**Comment by gbtitus** [comment]
After discussion with @ronawho, the alternatives for enabling comm=ofi testing on `srun`-based clusters seem to be:
- Launch with a Chapel GASNet launcher and use `gasnet_*()` functions for the out-of-band support.
- Launch with the Chapel MPI launcher and use `MPI_*()` functions for the out-of-band support. This requires augmenting our MPI launcher to support multilocale.
- Launch with the Chapel `srun` launcher.  This requires augmenting the `srun` launcher and the comm=ofi out-of-band support to respectively do the server and client side of `all_to_all` using sockets.
- Launch with anything, and write a comm=ofi out-of-band module based on files in a shared filesystem.  This works only when the program instances can all see a shared filesystem, but that's true in any of our internal testing scenarios.  This would only touch the comm=ofi out-of-band code.

The first two of these require linking comm=ofi programs with other communication libraries (GASNet or MPI, respectively), which is pretty heavyweight when all we really need is `all_to_all`. There is probably also some potential for unexpected interference with listener threads or other activities within those libraries. The third (`srun`) option seems the most elegant, but @ronawho has argued that it wouldn't have any purpose outside our own testing because anybody inclined to run multilocale Chapel would be using `mpirun` instead of `srun`. So maybe elegance isn't of value here. That argument also applies to the last (files) option. It's certainly a hack but it's definitely also easy, and maybe a hack is what's called for in this case.

@ronawho and @bradcray, I'd take votes on these alternatives. Potentially biasing you, I'll say up front that I do not like the first two.

**Comment by gbtitus** [comment]
One more minor note regarding the last two alternatives [here](https://github.com/chapel-lang/chapel/issues/11116#issuecomment-446768870): If it wasn't obvious from my description, the last (file-based) solution would be specific to our own core team environment, for testing and development. I don't see it leading or contributing to a generally available capability. The third (`srun` and sockets) solution would be more likely to find broader use, although note @ronawho's expectation that `mpirun` is more commonly used than `srun`.

**Comment by ronawho** [comment]
> Launch with the Chapel MPI launcher and use MPI_*() functions for the out-of-band support. This requires augmenting our MPI launcher to support multilocale.

> ... require linking comm=ofi programs with other communication libraries (GASNet or MPI, respectively), which is pretty heavyweight when all we really need is all_to_all. There is probably also some potential for unexpected interference with listener threads or other activities within those libraries. 

I think it'd be worth checking with the gasnet team to see how much overhead/interference we might expect from that.

If there isn't much interference using mpirun and mpi calls to bootstrap seems like it should be simple, scalable and portable.

**Comment by gbtitus** [comment]
After extensive back-and-forth with @ronawho, I'm now thinking I'll throw together a prototype of an ofi+mpirun launcher, probably written in python, just to see if any major issues arise and also to unblock #11856.  That unblock is my major short-term goal, and doing a simplistic `mpirun`-based launcher to achieve it shouldn't take any longer than either of my original preferred alternatives, while at the same time offering more future value than either of those.

**Comment by gbtitus** [comment]
I have a _very_ preliminary version of an `mpirun4ofi` running now, so I'm unblocking #11856.

**Comment by gbtitus** [comment]
It turns out the OpenMPI `mpirun` seems to work adequately for this, on our `chapcs` cluster.  When run within a slurm `salloc` job it automatically uses the slurm job's allocated nodes.  I modified the existing `mpirun` launcher to create an `mpirun4ofi` one, as follows:
- made it work only for comm=ofi instead of comm=none
- passed the Chapel launcher `-nl` option value to `mpirun`'s `-np` option
- threw `-bind-to none:overload-allowed` to `mpirun`, as a hack to allow testing with more Chapel locales than slurm job nodes and/or to allow multiple multilocale Chapel programs to run at the same time within the slurm job.

I also tried the `mpich` module on `chapcs` but was unable to get that to work; I saw a number of different kinds of internal errors when I tried it.  This may well have been user error, but the error messages certainly didn't give me any direction as to what to change or do, so I didn't pursue this any further.
